AX: Default to walking the DOM rather than the render tree when building the accessibility tree

twilco opened the pull request:
#### 6a44b38c10d068df677abf94c3f1d96e75fefb7a
<pre>
AX: Default to walking the DOM rather than the render tree when building the accessibility tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=281749">https://bugs.webkit.org/show_bug.cgi?id=281749</a>
<a href="https://rdar.apple.com/138184652">rdar://138184652</a>

Reviewed by Chris Fleizach.

Walking the DOM rather than the render tree allows many simplifications to our code. Before this
patch, we had several display:contents special cases scattered throughout the code that are no longer
needed:

  1. AccessibilityRenderObject::addNodeOnlyChildren() added display:contents children to the right
     place in an object&apos;s render-tree children retroactively. This function is now unnecessary —
     we get this property for free by walking the DOM to build the AX tree.

  2. AccessibilityObject::insertChild had a special case to prevent double-insertion of display:contents children.

  3. AccessibilityRenderObject::addCanvasChildren() only existed because canvas descendants are not rendered,
     and thus not picked up by a render tree walk. This function is now removed.

  4. Numerous other more minor checks throughout the code that only existed because we had to deal with
     anonymous renderers are removed.

The flip side of this is that we don&apos;t get &quot;interesting&quot; anonymous renderers for free anymore, e.g.
list markers, CSS ::before and CSS ::after, mfenced element `open`, `close`, `separators` anonymous
renderers. This is still a great tradeoff, as there are far more uninteresting anonymous renderers
than there are interesting ones.

I tried to apply this change to USE(ATSPI) ports, but there are GTK tests that expect the presence of certain
anonymous renderers that I&apos;m not sure we can outright remove (e.g. those in accessibility/gtk/replaced-objects-in-anonymous-blocks.html).
I filed a follow-up bug (<a href="https://bugs.webkit.org/show_bug.cgi?id=282117)">https://bugs.webkit.org/show_bug.cgi?id=282117)</a> to track this, more details are there.

Another important aspect of this change is the addition of AXObjectCache::onStyleChange. We have accessibility
properties cached based on element styles that never get updated, and the addition of this function gives us the
framework for starting to fix those.

* LayoutTests/accessibility/mac/ruby-hierarchy-roles-expected.txt:
* LayoutTests/accessibility/mac/ruby-hierarchy-roles.html:
Remove an else block from this test that only existed due to a useless
anonymous renderer child.
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::getOrCreate):
(WebCore::AXObjectCache::handleChildrenChanged):
(WebCore::AXObjectCache::onStyleChange):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityMathMLElement.cpp:
(WebCore::AccessibilityMathMLElement::addChildren):
* Source/WebCore/accessibility/AccessibilityMathMLElement.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::parentObject const):
(WebCore::AccessibilityNodeObject::addChildren):
(WebCore::shouldUseAccessibilityObjectInnerText):
(WebCore::AccessibilityNodeObject::parentObjectIfExists const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::insertChild):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::parentInCoreTree const):
(WebCore::AccessibilityObject::parentObjectIfExists const): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::parentObject const):
(WebCore::AccessibilityRenderObject::textUnderElement const):
(WebCore::AccessibilityRenderObject::shouldGetTextFromNode const):
(WebCore::AccessibilityRenderObject::boundingBoxRect const):
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
(WebCore::AccessibilityRenderObject::markerRenderer const):
(WebCore::AccessibilityRenderObject::addListItemMarker):
(WebCore::AccessibilityRenderObject::addChildren):
(WebCore::AccessibilityRenderObject::parentObjectIfExists const): Deleted.
(WebCore::AccessibilityRenderObject::addCanvasChildren): Deleted.
(WebCore::AccessibilityRenderObject::addNodeOnlyChildren): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/AccessibilityScrollView.h:
* Source/WebCore/accessibility/AccessibilityTable.cpp:
(WebCore::AccessibilityTable::addChildren):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper detailParentForSummaryObject:]):
(-[WebAccessibilityObjectWrapper accessibilitySupportsARIAExpanded]):
(-[WebAccessibilityObjectWrapper accessibilityIsExpanded]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateChildren):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolveComposedTree):

Canonical link: <a href="https://commits.webkit.org/285807@main">https://commits.webkit.org/285807@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/59aa64c5c827b58782d79c578805bdd6830854ce

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73613 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53042 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26424 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77925 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24851 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75728 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62175 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/827 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57885 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16277 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76680 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48037 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63341 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38291 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44628 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20827 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23184 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66366 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21176 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79501 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/930 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/406 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66241 "Found 2 new test failures: imported/blink/compositing/animation/hidden-animated-layer-should-not-have-scrollbars.html imported/w3c/web-platform-tests/css/css-animations/translation-animation-subpixel-offset.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1072 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63351 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65522 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16258 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9382 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7564 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/894 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3644 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/923 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/910 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/929 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/mac/attributed-string-with-listitem-multiple-lines.html:
```diff
@@ -18,7 +18,7 @@
     if (window.accessibilityController) {
         var listItem1 = accessibilityController.accessibleElementById("item1");
-        var p1 = listItem1.childAtIndex(0);
+        var p1 = listItem1.childAtIndex(1);
```
Prior to this PR, we sometimes used to insert list markers as not the first child of list items, which is wrong. I fixed this bug as a side effect of this change, and thus the `<p>` is at index 1 (as it should be) rather than index 0.

twilco commented on LayoutTests/accessibility/mac/ruby-hierarchy-roles.html:
```diff
@@ -41,12 +41,6 @@
         output += expect("axRuby.role", "expectedRubyRole");
         if (expectInline)
             output += expect("axRuby.subrole", "expectedRubyInlineSubrole");
-        else {
-            axRuby = axRuby.childAtIndex(0);
-            logId(axRuby);
-            output += expect("axRuby.role", "expectedRubyRole");
-            output += expect("axRuby.subrole", "expectedRubyInlineSubrole");
-        }
```
This branch only existed because we had to deal with anonymous renderers.

fleizach commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -856,7 +874,8 @@ AccessibilityObject* AXObjectCache::getOrCreate(Node& node, IsPartOfRelation isP
         return getOrCreate(*renderer);
 
-    if (!node.parentElement())
+    RefPtr composedParent = node.parentElementInComposedTree();
```
can we make this a reference here instead of dereferencing later on?

twilco replied:
I don't think so, unfortunately, since this can be null. We could turn it into a `Ref` after null-checking, but there's not much difference between a `Ref` and a `RefPtr` that has been checked for null. Both hold a strong reference count and thus guarantee the pointed to object is valid for as long as the scope. No ergonomic difference either — both require usage of `->` to get at the underlying value.

fleizach reviewed (approved):
LGTM, only minor comments

twilco committed:
AX: Default to walking the DOM rather than the render tree when building the accessibility tree
https://bugs.webkit.org/show_bug.cgi?id=281749
rdar://138184652

Reviewed by Chris Fleizach.

Walking the DOM rather than the render tree allows many simplifications to our code. Before this
patch, we had several display:contents special cases scattered throughout the code that are no longer
needed:

  1. AccessibilityRenderObject::addNodeOnlyChildren() added display:contents children to the right
     place in an object's render-tree children retroactively. This function is now unnecessary —
     we get this property for free by walking the DOM to build the AX tree.

  2. AccessibilityObject::insertChild had a special case to prevent double-insertion of display:contents children.

  3. AccessibilityRenderObject::addCanvasChildren() only existed because canvas descendants are not rendered,
     and thus not picked up by a render tree walk. This function is now removed.

  4. Numerous other more minor checks throughout the code that only existed because we had to deal with
     anonymous renderers are removed.

The flip side of this is that we don't get "interesting" anonymous renderers for free anymore, e.g.
list markers, CSS ::before and CSS ::after, mfenced element `open`, `close`, `separators` anonymous
renderers. This is still a great tradeoff, as there are far more uninteresting anonymous renderers
than there are interesting ones.

I tried to apply this change to USE(ATSPI) ports, but there are GTK tests that expect the presence of certain
anonymous renderers that I'm not sure we can outright remove (e.g. those in accessibility/gtk/replaced-objects-in-anonymous-blocks.html).
I filed a follow-up bug (https://bugs.webkit.org/show_bug.cgi?id=282117) to track this, more details are there.

Another important aspect of this change is the addition of AXObjectCache::onStyleChange. We have accessibility
properties cached based on element styles that never get updated, and the addition of this function gives us the
framework for starting to fix those.

* LayoutTests/accessibility/mac/ruby-hierarchy-roles-expected.txt:
* LayoutTests/accessibility/mac/ruby-hierarchy-roles.html:
Remove an else block from this test that only existed due to a useless
anonymous renderer child.
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::getOrCreate):
(WebCore::AXObjectCache::handleChildrenChanged):
(WebCore::AXObjectCache::onStyleChange):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityMathMLElement.cpp:
(WebCore::AccessibilityMathMLElement::addChildren):
* Source/WebCore/accessibility/AccessibilityMathMLElement.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::parentObject const):
(WebCore::AccessibilityNodeObject::addChildren):
(WebCore::shouldUseAccessibilityObjectInnerText):
(WebCore::AccessibilityNodeObject::parentObjectIfExists const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::insertChild):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::parentInCoreTree const):
(WebCore::AccessibilityObject::parentObjectIfExists const): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::parentObject const):
(WebCore::AccessibilityRenderObject::textUnderElement const):
(WebCore::AccessibilityRenderObject::shouldGetTextFromNode const):
(WebCore::AccessibilityRenderObject::boundingBoxRect const):
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
(WebCore::AccessibilityRenderObject::markerRenderer const):
(WebCore::AccessibilityRenderObject::addListItemMarker):
(WebCore::AccessibilityRenderObject::addChildren):
(WebCore::AccessibilityRenderObject::parentObjectIfExists const): Deleted.
(WebCore::AccessibilityRenderObject::addCanvasChildren): Deleted.
(WebCore::AccessibilityRenderObject::addNodeOnlyChildren): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/AccessibilityScrollView.h:
* Source/WebCore/accessibility/AccessibilityTable.cpp:
(WebCore::AccessibilityTable::addChildren):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper detailParentForSummaryObject:]):
(-[WebAccessibilityObjectWrapper accessibilitySupportsARIAExpanded]):
(-[WebAccessibilityObjectWrapper accessibilityIsExpanded]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateChildren):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolveComposedTree):

Canonical link: https://commits.webkit.org/285807@main
